### PR TITLE
server: fix semaphore release

### PR DESCRIPTION
**Pull request by bboreham**
This is to fix a bug in v1.57.1 where servers will run very sluggishly, and also take ages to shut down.

The acquire and release functions should have opposite impacts on the count.

Credit to @colega who spotted the diff.

This was introduced in #6706, a backport of #6703, where the line is correct.
https://github.com/grpc/grpc-go/pull/6703/files#diff-366e46a40f6f60b4f7614eb0976bb51820364bf5ca6ccc4787eb49d7bdbef3e6R2098

Also correct in the 1.56 backport: https://github.com/grpc/grpc-go/pull/6708/files#diff-366e46a40f6f60b4f7614eb0976bb51820364bf5ca6ccc4787eb49d7bdbef3e6R2073

RELEASE NOTES: n/a

### CODE REVIEW

**Review by dfawley — APPROVED** [review]
:eyes: Not sure how this happened, sorry!

**Comment by dfawley** [comment]
> I looked at the CI failure; can’t see how it can be related to this PR.

It isn't related, and it's an optional checker so won't block merging.  However, we do require 2 maintainers to review PRs before merging, which is why I haven't merged it yet.

**Review by ginayeh — APPROVED** [review]

**Pull request merged**
